---
title: "Transcriptomics and the analysis of RNA-Seq data"
output: github_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

#DESeq2 setup

We will use the DESeq2 package from bioconductor. To install this we ran **in the CONSOLE** - note eval=FALSE option!:
```{r eval=FALSE}
install.packages("BiocManager")
BiocManager::install()
BiocManager::install("DESeq2")
```

##Input data

```{r}
counts <- read.csv("airway_scaledcounts.csv", stringsAsFactors = FALSE)
metadata <-  read.csv("airway_metadata.csv", stringsAsFactors = FALSE)
```
Have a peak...
```{r}
head(counts)
```

```{r}
head(metadata)
```

How many genes do we have in this dataset?
```{r}
nrow(counts)
```

How many experiments do we have 
```{r}
ncol(counts)
```


```{r}
nrow(metadata)
```

3. Toy differential gene expression
we want to know if there is a differnece in expression values for contraol (non-drug) vs treated (i.e drug added to the cell lines)

First step is to find the mean value of the control and the treated and compare them.
```{r}
View(metadata)
```

We're going to specify the "dex" using the "=="
```{r}
control <- metadata[metadata[,"dex"]=="control",]
control.mean <- rowSums( counts[ ,control$id] )/4 
names(control.mean) <- counts$ensgene
```

Why did we divide by 4?
```{r}
control <- metadata[metadata[,"dex"]=="control",]
control$id
```

Q1. How would you make the above code more robust? It is not robust bc the code is hard coded to divide by four. This would not be favorable if the data set changed.
What would happen if you were to add more samples. Would the values obtained with the excat code above be correct?

How calculate the mean calues accross these control columns of **countsdata**
```{r}
control.mean<- rowSums(counts[,control$id]/length(control$id))
names(control.mean)<- counts$ensgene
control.mean
```


Do the same for the treated 
```{r}
treated <- metadata[metadata[,"dex"]=="treated",]
treated.mean <- rowSums( counts[ ,treated$id] )/4 
names(control.mean) <- counts$ensgene
treated.mean<- rowSums(counts[,treated$id]/length(treated$id))
names(treated.mean)<- counts$ensgene
```

Calculated the sum of the the averages
```{r}
meancounts <- data.frame(control.mean, treated.mean)
colSums(meancounts)
```

Plot control vs treated!

```{r}
plot(meancounts$control.mean, meancounts$treated.mean)
```


How to read the dots on the graphs above. if there is a diagonal live and the point lies above the the lie the expression is increase with the treated cells and if the point is below the line the expression is decreased in the treated cells


Created the log graph because the other graph was too large to see the trend
```{r}
plot(meancounts$control.mean, meancounts$treated.mean, log="xy")
```


We can find candidate differentially expressed genes by looking for genes with a large change between control and dex-treated samples. We usually look at the log2 of the fold change, because this has better mathematical properties.

Here we calculate log2foldchange, add it to our meancounts data.frame and inspect the results either with the head() or the View() function for example
```{r}
meancounts$log2fc <- log2(meancounts[,"treated.mean"]/meancounts[,"control.mean"])
head(meancounts)
```
You can't talk about the ones with 0.00 in their data bc it means that you have no stat to compare

There are a couple of “weird” results. Namely, the NaN (“not a number”“) and -Inf (negative infinity) results.

The NaN is returned when you divide by zero and try to take the log. The -Inf is returned when you try to take the log of zero. It turns out that there are a lot of genes with zero expression. Let’s filter our data to remove these genes. Again inspect your result (and the intermediate steps) to see if things make sense to you

```{r}
zero.vals <- which(meancounts[,1:2]==0, arr.ind=TRUE)

to.rm <- unique(zero.vals[,1])
mycounts <- meancounts[-to.rm,]
head(mycounts)
```
#which(meancounts[,1:2]==0 helps id the position in which the values in the column 1:2 that are = 0

remember 
x<- 56,0,26,0
x==0 
 will tell you true or false
 
which(x=-0)
will tell where the value is == 0

#the which (... arr.ind) will tell the  positions of the zeros in the data frame
```{r}
x <-data.frame(control=c(56,0,5), treated=c(26,0,0))
x
which(x==0, arr.ind = TRUE)
```

A common threshold used for calling something differentially expressed is a log2(FoldChange) of greater than 2 or less than -2. Let’s filter the dataset both ways to see how many genes are up or down-regulated.

```{r}
up.ind <- mycounts$log2fc > 2
down.ind <- mycounts$log2fc < (-2)
```


```{r}
sum(up.ind)
```



5. DESeq2 analysis
Let’s do this the right way. DESeq2 is an R package for analyzing count-based NGS data like RNA-seq. It is available from Bioconductor. Bioconductor is a project to provide tools for analyzing high-throughput genomic data including RNA-seq, ChIP-seq and arrays.
```{r}
library(DESeq2)
citation("DESeq2")
```


```{r}
dds <- DESeqDataSetFromMatrix(countData=counts, 
                              colData=metadata, 
                              design=~dex, 
                              tidy=TRUE)
dds
```



```{r}
sizeFactors(dds)
dispersions(dds)
dds <- DESeq(dds)
```
```{r}
results(dds)
```


```{r}
res <- results(dds)
res
```

```{r}
summary(res)
```


```{r}
resOrdered <- res[order(res$pvalue),]
```

```{r}
res05 <- results(dds, alpha=0.05)
summary(res05)
```


```{r}
resSig05 <- subset(as.data.frame(res), padj < 0.05)
nrow(resSig05)
```




6. Data Visualization
Plotting counts
DESeq2 offers a function called plotCounts() that takes a DESeqDataSet that has been run through the pipeline, the name of a gene, and the name of the variable in the colData that you’re interested in, and plots those values. See the help for ?plotCounts. Let’s first see what the gene ID is for the CRISPLD2 gene using:


```{r}
#add some colors
mycols<- rep("gray", nrow(res))
#make points with +2 -2 fold change blue
mycols[abs(res$log2FoldChange)>2]<-"blue"

#make points below p-value cutoff gray
mycols[(res$padj>0.050)]<- "gray"

plot( res$log2FoldChange,  -log(res$padj), col=mycols)
abline(v=c(-2,2),col="gray", lty=2)
abline(h=-log(0.05), col="gray", lty=2)
```




save our results
```{r}
write.csv(res, file="expression_reslts.csv")
```

```{r}
view(expression_results.csv)
```








