
title: "Untitled"
output: github_document
---

```{r}
library(GenomicDataCommons)
library(TCGAbiolinks)
library(maftools)
```

```{r}
status()
```

```{r}
projects <- getGDCprojects()
head(projects)
```

```{r}
cases_by_project <- cases() %>%
  facet("project.project_id") %>%
  aggregations()
head(cases_by_project)
```
Q9. Write the R code to make a barplot of the cases per project. Lets plot this data with a log scale for the y axis (log="y"), rotated axis labels (las=2) and color the bar coresponding to the TCGA-PAAD project.

```{r}
x <- cases_by_project$project.project_id

# Make a custom color vector for our plot
colvec <- rep("lightblue", nrow(x))
colvec[(x$key == "TCGA-PAAD")]<- "red"

# Plot with 'log' for y axis and rotate labels with 'las'
#par(___)  
barplot( x$doc_count , names.arg= x$key, log="y", col=colvec, las=2)
```


```{r}
#make custom colors
colvec <-rep("lightblue", nrow(x))
colvec[ (x$key == "TCGA-PAAD")] <- "red"
#colvec

barplot(x$doc_count, col=colvec, log="y")
```



```{r}
samp <- getSampleFilesSummary("TCGA-PAAD")
head(samp)
```

```{r}
query <- GDCquery(project="TCGA-PAAD",
                  data.category="Transcriptome Profiling",
                  data.type="Gene Expression Quantification")

ans <- getResults(query)
```


```{r}
head(ans)

```


```{r}
View(ans)
```

```{r}
nrow(ans)
```

4. Variant analysis with R
Note we could go to the NCI-GDC web portal and enter the Advanced Search page and then construct a search query to find MAF format somatic mutation files for our ‘TCGA-PAAD’ project.

*Q9. How many MAF files for the TCGA-PAAD project were found from this advanced web search?*

```{r}
maf.file <- GDCquery_Maf(tumor="PAAD", pipelines = "mutect")
```
```{r}
head(maf.file)
```

*Q10. What argument could we use to write the MAF file into a csv document in your current working directory?*

```{r}
vars = read.maf(maf = maf.file, verbose = FALSE)

```

```{r}
plotmafSummary(vars)
```


```{r}
oncoplot(maf = vars, top = 10)
```

```{r}
# Oncoplot for our top 10 most frequently mutated genes
pdf("oncoplot_panc.pdf")
oncoplot(maf = vars, top = 10, fontSize = 12)
dev.off()
```

```{r}
oncostrip(maf=vars, genes=c("KRAS", "TP53"))

```


```{r}
lollipopPlot(vars, gene='KRAS')
```

Q11. Adapt the code above to produce a lollipop plot for p53 (i.e. the ‘TP53’ gene)?

```{r}
lollipopPlot(vars, gene='TP53')
```



part II

```{r}
library(bio3d)
seqs <- read.fasta("lecture18_sequences.fa")
seqs
```
align sequences
```{r}
seqs <- seqaln(seqs)
```


```{r}
## Calculate positional identity scores
ide <- conserv(seqs$ali, method="identity")
mutant.sites <- which(ide < 1) 

## Exclude gap possitions from analysis
gaps <- gap.inspect(seqs)
mutant.sites <- mutant.sites[mutant.sites %in% gaps$f.inds]

mutant.sites
```


```{r}
## Make a "names" label for our output sequences (one per mutant)
mutant.names <- paste0(seqs$ali["P53_wt",mutant.sites],
                       mutant.sites,
                       seqs$ali["P53_mutant",mutant.sites])

mutant.names
```

```{r}
## Sequence positions surounding each mutant site
start.position <- mutant.sites - 8
end.position <-  mutant.sites + 8

# Blank matrix to store sub-sequences
store.seqs <- matrix("-", nrow=length(mutant.sites), ncol=17)
rownames(store.seqs) <- mutant.names

## Extract each sub-sequence
for(i in 1:length(mutant.sites)) {
  store.seqs[i,] <- seqs$ali["P53_mutant",start.position[i]:end.position[i]]
}

store.seqs
```

```{r}
## Output a FASTA file for further analysis
write.fasta(seqs=store.seqs, ids=mutant.names, file="subsequences.fa")
```

```{r}
library(bio3d)

## Read KRas oncogene sequence from UniProt
wt <- get.seq("P01116")

## Here we make four mutants namely: G12V, Q22N, T74S and A130V
mutant <- wt
mutant$ali[ c(12,22,74,130)] <- c("V", "N", "S", "V")

write.fasta( seqbind(wt, mutant), ids=c("wt","mutant"), file="kras-sequences.fa")
```

```{r}
sessionInfo()
```





































